### PR TITLE
Parallelize initial Rust download in bootstrap

### PR DESCRIPTION
**Pull request by Noratrieb**
Parallelize the initial download of Rust in `bootstrap.py`

`time ./x.py --help` after `rm -r build`
Before: 33s
After: 27s

### CODE REVIEW

**Comment by Nilstrieb** [comment]
Apparently we want to be python 2 compatible here, I'll fix that tomorrow.

**Comment by jyn514** [comment]
I think we might be able to drop python2 compat in bootstrap.py proper now that the x.py shim reexecs itself with python3.

**Comment by Nilstrieb** [comment]
I wasn't able to see this on my fast internet connection, but testing this on a slower one shows that the progress bars get pretty broken by the parallel download which is not ideal.

**Comment by Nilstrieb** [comment]
I made it Python 2.7 compatible and tested it manually using `python2.7 x.py` (patching `x.py` to not exec Python 3).

**Comment by albertlarsan68** [comment]
Is the progress bar mess fixed?

**Comment by Nilstrieb** [comment]
Inserting an `exit()` before building bootstrap

with a fast (480MiB/s) internet connection
serial download, parallel unpacking: 11s
parallel download, parallel unpacking: 10s
serial download, serial unpacking: 16s

with a less fast (60MiB/s) internet connection
serial download, parallel unpacking: 16s
parallel download, parallel unpacking: 22s (the terminal output from the download progress bars is completely broken)
serial download, serial unpacking: 26s

so the parallel download is actually worse for slow connections, probably because it reduces parallelism for the extraction phase since the downloads don't all complete at the same time

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -429,17 +472,49 @@ def download_toolchain(self):
...
+                for filename, pattern in tarballs_to_download
+            ]
+
+            # Download the components serially to show the progress bars properly.
+            for download_info in tarballs_download_info:
+                download_component(download_info)
+
+            # Unpack the tarballs in parallle.
+            # In Python 2.7, Pool cannot be used as a context manager.
+            p = Pool(3)
```
Comment (albertlarsan68) [inline src/bootstrap/bootstrap.py:512]: Would it be possible to use something like `min(len(tarballs_to_download), available_parallelism())`?
I suspect this step is CPU-bound on lower-end CPUs, which are likely to have only 2 cores.

**Review by albertlarsan68 — COMMENTED** [review]
LGTM, just one question/nit/potential improvement.

**Comment by albertlarsan68** [comment]
Thanks for the PR!
@bors r+ rollup=iffy (touches core bootstrap code)

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -429,17 +472,49 @@ def download_toolchain(self):
...
+            for download_info in tarballs_download_info:
+                download_component(download_info)
+
+            # Unpack the tarballs in parallle.
+            # In Python 2.7, Pool cannot be used as a context manager.
+            p = Pool(min(len(tarballs_download_info), cpu_count()))
+            try:
+                p.map(unpack_component, tarballs_download_info)
+            finally:
+                p.close
```
Comment (alex) [inline src/bootstrap/bootstrap.py:516]:
```suggestion
                p.close()
```

**Comment by Nilstrieb** [comment]
I quickly tested it locally again and it worked.
@bors r=albertlarsan68

**Comment by Zalathar** [comment]
This seems to be causing bootstrap to get stuck in a loop and unable to proceed on some systems.

I'm seeing problems on macOS, and #111046 reports similar problems on Windows.

**Comment by Zalathar** [comment]
[It seems that on some platforms, `multiprocessing` *requires* the program's entry point to have been guarded by the `if __name__ == '__main__':` idiom.](https://docs.python.org/3/library/multiprocessing.html#multiprocessing-safe-main-import)

If I edit `x.py` to wrap all the code in that if statement, bootstrap seems to work fine again.

**Pull request merged**